[rpc] listsinceblock should include lost transactions when parameter is a reorg'd block

kallewoof opened the pull request:
The following scenario will not notify the caller of the fact `tx0` has been dropped:

1. User 1 receives BTC in tx0 from utxo1 in block aa1.
2. User 2 receives BTC in tx1 from utxo1 (same) in block bb1
3. User 1 sees 2 confirmations at block aa3.
4. Reorg into bb chain.
5. User 1 asks `listsinceblock aa3` and does not see that tx0 is now invalidated.

See `listsinceblock.py` commit for related test.

The proposed fix is to iterate from the given block down to the fork point, and to check each transaction in the blocks against the wallet, in addition to including all transactions from the fork point to the active chain tip (the current behavior). Any transactions that were present will now also be listed in the `listsinceblock` output in a new `replaced` array. This operation may be a bit heavy but the circumstances (and perceived frequency of occurrence) warrant it, I believe.

Example output:
```Python
{
  'transactions': [],
  'replaced': [
    {
      'walletconflicts': [],
      'vout': 1,
      'account': '',
      'timereceived': 1485234857,
      'time': 1485234857,
      'amount': '1.00000000',
      'bip125-replaceable': 'unknown',
      'trusted': False,
      'category': 'receive',
      'txid': 'ce673859a30dee1d2ebdb3c05f2eea7b1da54baf68f93bb8bfe37c5f09ed22ff',
      'address': 'miqEt4kWp9zSizwGGuUWLAmxEcTW9bFUnQ',
      'label': '',
      'confirmations': -7
    }
  ],
  'lastblock': '7a388f27d09e3699102a4ebf81597d974fc4c72093eeaa02adffbbf7527f6715'
}
```

I believe this addresses the comment by @luke-jr in https://github.com/bitcoin/bitcoin/pull/9516#issuecomment-274190081 but I could be wrong..

kallewoof committed:
listsinceblock: optionally find and list any transactions that were undone due to reorg when requesting a non-main chain block in a new 'removed' array.

kallewoof committed:
Testing: listsinceblock should display all transactions that were affected since the given block, including transactions that were removed due to a reorg.

jonasschnelli commented:
I'm not 100% sure if we want this. If we, we would need at least to updated the `listsinceblock`'s RPC help message to mention that if one requests tx's from a block outside the main-chain, that the result will also contain transactions not in the main chain.

kallewoof commented:
@jonasschnelli Updated the help output. I am not sure what the reasons for not wanting this are, unless you're referring to resource consumption. I think it's a rare (but important) enough case to warrant it.

It would definitely make it easier for RPC applications checking the validity of existing transactions to explicitly provide these when a reorg affects them. The only other alternative right now is to keep a list of transactions with confirmations less than some arbitrary number (100) and to loop through these every time a reorg is encountered to ensure they're actually still present.

jonasschnelli commented:
@kallewoof: I missed the point that if you pass in an orphan block to list since, you also get the transactions upwards the chain-fork on the main chain. At first sight, I though you get only tx from the re-orged-off chain in that case.

Concept ACK (and I think it would be clever to list them in an extra array element `reorged:[]`.

kallewoof commented:
@jonasschnelli Gotcha. I updated the OP to clarify that it's also including transactions from the fork point to the active chain tip. I also moved the off chain transactions into a new 'reorged' array. (f501acc & 461d5a3)

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1729,6 +1730,25 @@ UniValue listsinceblock(const JSONRPCRequest& request)
+        {
+            if (pwalletMain->mapWallet.count(tx->GetHash()))
+            {
+                ListTransactions(pwalletMain->mapWallet[tx->GetHash()], "*", -depth, true, transactions, filter);
```
I think it would be helpful to add a comment about -depth here. I was staring at this a long time to figure out how it worked. Comment could say: "Pass -depth as minDepth to prevent any filtering in ListTransactions. (Works because tx can only conflict with transactions after pindex, so GetDepthInMainChain will always return at least (1-depth))."

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1641,7 +1641,9 @@ UniValue listsinceblock(const JSONRPCRequest& request)
             "listsinceblock ( \"blockhash\" target_confirmations include_watchonly)\n"
-            "\nGet all transactions in blocks since block [blockhash], or all transactions if omitted\n"
+            "\nGet all transactions in blocks since block [blockhash], or all transactions if omitted.\n"
+            "If \"blockhash\" is no longer a part of the main chain, all transactions affecting the node wallet back to the fork point are included, as well as those from \n"
```
Would s/back to the fork point/from blockhash back to the fork point/ to clarify, because it sounds to me like this is referring to transactions between the active tip and the fork point (making the rest of the sentence confusing).

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1756,6 +1757,7 @@ UniValue listsinceblock(const JSONRPCRequest& request)
 
     UniValue ret(UniValue::VOBJ);
     ret.push_back(Pair("transactions", transactions));
+    ret.push_back(Pair("reorged", reorged));
```
Need documentation update to accompany this.

ryanofsky commented on qa/rpc-tests/listsinceblock.py:
```diff
@@ -76,5 +76,98 @@ def run_test (self):
+        privkey = self.nodes[2].dumpprivkey(utxo['address'])
+        self.nodes[1].importprivkey(privkey)
+
+        # send from nodes[2] using utxo to nodes[0]
```
Really sending from node 1  here (even though the originally key comes from node 2)

ryanofsky commented on qa/rpc-tests/listsinceblock.py:
```diff
@@ -76,5 +76,98 @@ def run_test (self):
+        self.join_network()
+
+        # gettransaction should work for txid1
+        tmp = self.nodes[0].gettransaction(txid1)
```
unused variable

ryanofsky commented on qa/rpc-tests/listsinceblock.py:
```diff
@@ -76,5 +76,98 @@ def run_test (self):
+
+        # listsinceblock(lastblockhash) should now include txid1, as seen from nodes[0]
+        lsbres = self.nodes[0].listsinceblock(lastblockhash)
+        found = False
```
Maybe condense this block to a single line, and add a check for txid2 (untested)

```
assert_equal(any(tx['txid'] == txid1 for tx in lsbres['reorged']), True)
assert_equal(any(tx['txid'] == txid2 for tx in lsbres['transactions']), True)
```

kallewoof replied:
@ryanofsky Ahh, nice! I didn't know about `any()`.

Since txid2 is not related to nodes[0], it will not list it anywhere so that second assert_equal will not be true.

Thanks for all the feedback! I believe everything you suggested is in 6953994 & 4faf084.

ryanofsky reviewed (approved):
utACK 461d5a37a3d83edbeedb701ed207bc14412dee0d with minor suggestions

ryanofsky commented on qa/rpc-tests/listsinceblock.py:
```diff
@@ -76,5 +76,93 @@ def run_test (self):
+
+             ab0
+          /       \
+        aa1 [tx0]   bb1 [tx1]
```
Maybe change tx0 and tx1 to tx1 and tx2 to match code. Also code is generating 6 and 7 blocks instead of 3 and 4, so maybe that could be changed to match the example.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1729,11 +1735,34 @@ UniValue listsinceblock(const JSONRPCRequest& request)
+            if (pwalletMain->mapWallet.count(tx->GetHash()))
+            {
+                // Use -depth as minDepth parameter to ListTransactions, as these transactions have
+                // negative (at minimum -depth) confirmations.
```
Would remove parentheses since (at minimum -depth) is the important part. Also, technically I believe the minimum is 1-depth. Maybe:

```
// Use -depth as minDepth parameter to ListTransactions to prevent incoming
// transactions from being filtered. These transactions have negative
// confirmations, but always greater than -depth.
```

kallewoof commented:
@ryanofsky Thanks for all the feedback! Updated.

gmaxwell commented:
Concept ACK.

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1729,11 +1735,35 @@ UniValue listsinceblock(const JSONRPCRequest& request)
+                // Use -depth as minDepth parameter to ListTransactions to prevent incoming
+                // transactions from being filtered. These transactions have negative
+                // confirmations, but always greater than -depth.
+                ListTransactions(pwalletMain->mapWallet[tx->GetHash()], "*", -depth, true, reorged, filter);
```
I dont believe this prevents listing transactions which were on both sides of the fork? I'm pretty sure we dont want to list such transactions in a "reorged" list.

TheBlueMatt reviewed (commented):
I'm not sure I'd call this a bugfix - we already correctly print transactions which were included in blocks from the common fork point from the blockhash provided to listsinceblock.

kallewoof commented:
@TheBlueMatt I guess it comes down to what the user expects. I personally expected listsinceblock to show me anything I needed to keep an updated score of what's going on since the given block. In that sense, not showing a transaction vanishing would probably be considered a bug.

Regarding showing an existing transaction in both reorged and transactions, I patched this by ensuring that any tx shown from the main chain is skipped over in the reorged list. See the new `test_double_send` test.

JeremyRubin commented:
I agree with Matt, this isn't a bugfix, it's feature :) 

I think I'm in favor of adding this. I would be more supportive of naming them `replaced` rather than reorged if I understand the semantics properly. Strictly speaking, I'd say a reorged txn is any transaction that was between the old tip and the fork point?

I would also think maybe making this information available with a default false parameter may be a good idea to introduce less changes for current users/legacy code.

laanwj commented:
Moving milestone to 0.15 as this was re-classified as a feature and the feature freeze is long past.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1640,12 +1657,15 @@ UniValue listsinceblock(const JSONRPCRequest& request)
             "1. \"blockhash\"            (string, optional) The block hash to list transactions since\n"
             "2. target_confirmations:    (numeric, optional) The confirmations required, must be 1 or more\n"
             "3. include_watchonly:       (bool, optional, default=false) Include transactions to watch-only addresses (see 'importaddress')"
+            "4. include_replaced:        (bool, optional, default=true) Show transactions that were replaced due to a reorg in the \"replaced\" array"
```
Not sure this needs to be optional...?

kallewoof replied:
Me neither. It was mostly suggested as a way for legacy code to cope better, but only really crappy legacy code (that broke when a new key was introduced) would be affected I guess.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1717,23 +1741,55 @@ UniValue listsinceblock(const JSONRPCRequest& request)
         if (depth == -1 || tx.GetDepthInMainChain() < depth)
-            ListTransactions(tx, "*", 0, true, transactions, filter);
+        {
+            if (ListTransactions(tx, "*", 0, true, transactions, filter) && include_replaced)
```
Why do we need `ListTransactions`'s return value? If it doesn't get inserted here, it won't later-on either, right?

kallewoof replied:
See above case with a tx being sent on both chains.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1334,12 +1334,26 @@ static void MaybePushAddress(UniValue & entry, const CTxDestination &dest)
+ * @param  filter     The "is mine" filter bool.
+ * @return            true if appends were made to ret, false otherwise.
+ */
+bool ListTransactions(const CWalletTx& wtx, const string& strAccount, int nMinDepth, bool fLong, UniValue& ret, const isminefilter& filter)
```
IMO this PR doesn't need to modify `ListTransactions`, and this ought to be move to a separate PR.

kallewoof replied:
For the case where tx0 is in reorged and main chain both (see `test_double_send`), both versions of the transaction will appear in the results (once in 'transactions' and once in 'replaced'). To prevent that, I check if `ListTransactions` appended to the transactions array -- if it did, I need to not include in replaced.

luke-jr reviewed (approved):
utACK as-is, with a few nits. (IMO this is a bug fix.)

luke-jr commented:
It occurs to me that the key `"replaced"` may not be ideal, since there is no guarantee these transactions are now conflicted and won't get re-mined. Perhaps `"removed"`?

kallewoof commented:
I tried to find actual cases where this happened, but had a hard time finding a case that wasn't the cause of a double spend. It would require that the transaction is NOT in the mempool of the node, or it would appear as normal in `transactions`. There may be a case I haven't thought of though, but you may be right, `removed` is a safer bet for sure.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1654,12 +1671,15 @@ UniValue listsinceblock(const JSONRPCRequest& request)
         throw runtime_error(
-            "listsinceblock ( \"blockhash\" target_confirmations include_watchonly)\n"
-            "\nGet all transactions in blocks since block [blockhash], or all transactions if omitted\n"
+            "listsinceblock ( \"blockhash\" target_confirmations include_watchonly include_replaced )\n"
```
"replaced"

kallewoof replied:
Done

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1654,12 +1671,15 @@ UniValue listsinceblock(const JSONRPCRequest& request)
-            "\nGet all transactions in blocks since block [blockhash], or all transactions if omitted\n"
+            "listsinceblock ( \"blockhash\" target_confirmations include_watchonly include_replaced )\n"
+            "\nGet all transactions in blocks since block [blockhash], or all transactions if omitted.\n"
+            "If \"blockhash\" is no longer a part of the main chain, all transactions affecting the node wallet from blockhash back to the fork point are included in the \"replaced\" array.\n"
```
"replaced"

kallewoof replied:
Done

luke-jr reviewed (changes requested):
Two more "replaced" instances. Also need to update the named param list at the bottom of the file, and in rpc/client.cpp

kallewoof commented:
@luke-jr Sorry about sloppiness, I thought I got all of them.

ryanofsky commented:
utACK ad57cefb6583ab669d8da21d0bf808e2d38ef04a.

Comparing against previously ACKed 461d5a37a3d83edbeedb701ed207bc14412dee0d, the both-sides dedup code and the new API tweaks and comments looked good.

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1654,12 +1671,15 @@ UniValue listsinceblock(const JSONRPCRequest& request)
-            "\nGet all transactions in blocks since block [blockhash], or all transactions if omitted\n"
+            "listsinceblock ( \"blockhash\" target_confirmations include_watchonly include_removed )\n"
+            "\nGet all transactions in blocks since block [blockhash], or all transactions if omitted.\n"
+            "If \"blockhash\" is no longer a part of the main chain, all transactions affecting the node wallet from blockhash back to the fork point are included in the \"removed\" array.\n"
```
Maybe instead of this and the next line,
"If \"blockhash\" is no longer a part of the main chain, transactions from the fork point onward are included.\n"
"Additionally, if include_removed is set, transactions affecting the wallet which were removed are returned in the \"removed\" array.\n"

kallewoof replied:
Makes sense. Changed.

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1731,23 +1755,55 @@ UniValue listsinceblock(const JSONRPCRequest& request)
+        CBlock block;
+        if (!ReadBlockFromDisk(block, paltindex, Params().GetConsensus()))
+        {
+            throw JSONRPCError(RPC_INTERNAL_ERROR, "Can't read block from disk");
```
Hmm, not sure how I feel about this. For pruning nodes, it may be more useful to return a partial list and a message indicating the results are incomplete rather than throw wholesale.

kallewoof replied:
Isn't that slightly dangerous? I mean, if people don't realize the node only has a subset of all blocks in some cases, they might miss the "partial" flag and never realize they potentially missed something. Even if it's their own node. Someone could even time an attack based on that.

To take a step back, the situation here is that some software which uses listsinceblock believes the chain is on one chain when it is on another one. Presuming this fork is long, the node ends up pruning beyond the fork point. And/or, the node ends up pruning out the alt chain that was deactivated. It feels like the caller (requester) should blow up at this point and either require manual assistance, or fall back to scanning over all tx's, effectively "resetting" to the current chain tip.

TheBlueMatt replied:
True. I'm not sure what the right answer here is...clearly we dont want it to be so easy to miss transactions, but on the other hand, it sucks to not have the option here. A different approach might be to iterate over mapWallet and find all transactions which have a hash of a disconnected block, which I believe is correct as long as the block was, at some point, on the active chain, though might not be if the user is calling listsinceblock on a block that was never the active chain.

kallewoof replied:
Hm.. It almost feels like I can just get the block headers by iterating `altchainindex` up to the fork point and put those in a set and find the removed transactions in the initial mapWallet iteration. That way there's no need to read in the blocks at all and can get rid of `listed` as well. Will try that, thanks for the nudge!

TheBlueMatt replied:
Yea, I was heasitant to recommend doing it alone because I'm not confident about how safe it is to make sure the wallet's data is always right, will have to think more and review once you have it coded :).

kallewoof replied:
It seems reasonable, but I'm a bit stuck on how to get the block hash for a tx that is not in the main chain. It seems like CMerkleTx's blockHash uint256 is only set if in the main chain...

TheBlueMatt replied:
Oh, sorry, you're right...hashBlock is always something in the main chain (so you would only be able to print conflicted transactions and not transactions which went back to 0conf).

kallewoof replied:
Options:
①. Try to respect the path taken which requires the blocks as removed transactions do not track the block they were in, meaning if a node prunes out the block, the operation fails.
②. Ignore the path and simply show all removed transactions between the fork point and the chain tip, which requires NO loading of blocks (faster), but means the caller will get irrelevant removed transactions in reorgs it was never aware of to begin with.

Pros and cons with ①:
1. Pro: Caller gets exactly what they need to update their view of the chain state, and is not confused by irrelevant stuff.
2. Con: Slower due to block read requirement.
3. Con: Fails if blocks involved in the reorg were pruned out.

Pros and cons with ②:
1. Pro: Faster due to no block read.
2. Pro: Works fine even if involved blocks were pruned.
3. Con: Does not distinguish between the various forks involved.

TheBlueMatt replied:
There is a second con to ② - it cannot list transactions which went from confirmed to 0-conf, as they now have a hashBlock of null, which I think is somewhat of a blocker to using that approach. Another option is to have an optional parameter (default to off) which allows you to get partial-removed-data.

kallewoof replied:
A 0-conf tx will be in the mempool right? Presuming it is, it will in fact be in the transactions array.

The allow partial parameter makes sense to me. I think I'll do that for starters until we make a decision (which may be another PR).

TheBlueMatt replied:
That isnt a guarantee, because mempool may be limited.

kallewoof replied:
@TheBlueMatt I took the liberty to remove he allow_partial flag as I've seen more support for always throwing than for having the flag and only throwing sometimes. The unsquashed tree has everything in it still so it'd be no effort to resurrect this, if you or someone felt strongly enough about it.

sipa replied:
I agree with (1), and not having `allow_partial`. We may want to document that this feature stops working with pruning.

TheBlueMatt reviewed (commented):
utACK ad57cefb6583ab669d8da21d0bf808e2d38ef04a +/- some slight API nitpicks.

kallewoof commented:
@TheBlueMatt I didn't think about that. 

But would the wallet actually do that to its own transaction? Throw out of mempool I mean.

kallewoof commented:
s/wallet/node/

sipa commented:
The wallet doesn't control the mempool, and the mempool behaves
independently. Anything else would be a privacy leak, as the mempool is
observable through the P2P network.

luke-jr commented:
IMO if the stale block is pruned, throwing an exception is the right thing to do. The caller should be able to figure out its own rewinding back to the last common block.

(P.S. Perhaps the tests here should cover the case where there are three potential-chains involved?)

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1734,23 +1762,62 @@ UniValue listsinceblock(const JSONRPCRequest& request)
+                throw JSONRPCError(RPC_INTERNAL_ERROR, "Can't read block from disk");
+            }
+        }
+        else for (const CTransactionRef& tx : block.vtx)
```
No `else for` please.

```C++
} else {
    for (const CTransactionRef& tx : block.vtx) {
```

kallewoof commented:
Good point about testing 3 chains. Will add a test like that.

kallewoof commented:
I just realized the split network feature in the QA framework only supports one split (2 chains) right now. I think I'll make a separate PR to extend that functionality and add more tests for 3+ chains after this is merged.

kallewoof commented:
Squashed commits. Unsquashed branch preserved at https://github.com/kallewoof/bitcoin/tree/listsinceblock-include-lost-txs-unsquashed for comparison.

MarcoFalke commented:
> I just realized the split network feature in the QA framework only
supports one split (2 chains) right now. I think I'll make a separate PR to
extend that functionality and add more tests for 3+ chains after this is
merged.

Yes, please. Also make sure the extended functionality does not depend on
the number of nodes in regtest being exactly 4.

On Sun, Feb 26, 2017 at 2:10 AM, kallewoof <notifications@github.com> wrote:

> Squashed commits. Unsquashed branch preserved at
> https://github.com/kallewoof/bitcoin/tree/listsinceblock-
> include-lost-txs-unsquashed for comparison.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/9622#issuecomment-282524695>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGGmv98uWx_PRJaQ16WpuPIQGTvhEo8Wks5rgNEBgaJpZM4Lr4lC>
> .
>

kallewoof commented:
@MarcoFalke See #9872.

jtimon commented:
Concept ACK. Agree with @luke-jr in throwing the exception.

kallewoof commented:
Rebased.

jnewbery commented:
I've only reviewed the test code. It's a shame that this makes the listsinceblock test runtime increase (from 32s to 72s on recent travis runs), although #10198 and  #10082 will help with that.

One design comment: I'm not convinced it's a good idea to *not* including duplicate transactions in the `removed` array. I think that could be a trap for users. My expectation would be that I could periodically run this RPC to get the list of transactions since last time I ran the command (by passing in the `lastblock` value from the previous call). I'd expect to be able to keep a set of transactions, adding the `transactions` array to my set and subtracting the `removed` array. If you don't include the duplicate transactions in the `removed` array, then the user might end up with more than one instance of a transaction in their set.

Obviously the user should look out for duplicates and not add them to their set, but if they're doing something daft like keying off the blockhash and txid, then they'll get into this inconsistent state.

I don't think that's enough to block this PR, but I think it's considering.

kallewoof commented on test/functional/listsinceblock.py:
```diff
@@ -77,5 +77,183 @@ def run_test (self):
+            'txid': utxo['txid'],
+            'vout': utxo['vout'],
+        }]
+        txid1 = self.nodes[1].sendrawtransaction(
```
Unfortunately `sendtoaddress` doesn't let me choose the UTXO so I don't think I can use it here.

kallewoof commented on test/functional/listsinceblock.py:
```diff
@@ -77,5 +77,183 @@ def run_test (self):
+        # generate on both sides
+        lastblockhash = self.nodes[1].generate(3)[2]
+        self.nodes[2].generate(4)
+        print('lastblockhash=%s' % (lastblockhash))
```
Removed all `print()`s. They seemed vaguely useful but not enough to warrant keeping them around long term.

kallewoof commented on test/functional/listsinceblock.py:
```diff
@@ -77,5 +77,183 @@ def run_test (self):
+        self.join_network()
+
+        # gettransaction should work for txid1
+        self.nodes[0].gettransaction(txid1)
```
If the transaction cannot be gotten, an exception is raised (I swapped the txid out for a random hex below):
```
2017-04-13 02:22:05.191000 TestFramework (ERROR): JSONRPC error
Traceback (most recent call last):
  File "/Users/karljohan-alm/workspace/bitcoin-kw/test/functional/test_framework/test_framework.py", line 148, in main
    self.run_test()
  File "./listsinceblock.py", line 250, in run_test
    self.test_double_spend()
  File "./listsinceblock.py", line 155, in test_double_spend
    self.nodes[0].gettransaction("000000000009f75870f0ebe6f317af6928e3a74c676a60a0f11d3eab51c88ff3")
  File "/Users/karljohan-alm/workspace/bitcoin-kw/test/functional/test_framework/coverage.py", line 46, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/Users/karljohan-alm/workspace/bitcoin-kw/test/functional/test_framework/authproxy.py", line 153, in __call__
    raise JSONRPCException(response['error'])
test_framework.authproxy.JSONRPCException: Invalid or non-wallet transaction id (-5)
2017-04-13 02:22:05.193000 TestFramework (INFO): Stopping nodes
2017-04-13 02:22:14.337000 TestFramework (WARNING): Not cleaning up dir /var/folders/b8/znqr8hj918772gfmd875gzgdd3ypz1/T/test5fli8q5i/13217
2017-04-13 02:22:14.337000 TestFramework (ERROR): Test failed. Test logging available at /var/folders/b8/znqr8hj918772gfmd875gzgdd3ypz1/T/test5fli8q5i/13217/test_framework.log
```
Asserting on the returned value is thus not necessary but I'm adding a comment indicating this.

jnewbery replied:
This comment is fine. Alternatively you could assert something like:

```python
assert self.nodes[0].gettransaction(txid1)['txid'] == txid1, "gettransaction failed to find txid1 not found"
```

kallewoof replied:
Ohh, I didn't realize assert also captured exceptions. Thanks, changed.

kallewoof commented on test/functional/listsinceblock.py:
```diff
@@ -77,5 +77,183 @@ def run_test (self):
+            if tx['txid'] == txid1:
+                assert_equal(tx['confirmations'], 2)
+
+    def run_test(self):
```
Good idea. Changing.

kallewoof commented:
@jnewbery Thanks a lot for the review! I've addressed most of your concerns, except for the nit about testing allow_partial flag.

Regarding duplicates in `removed`, I'm neutral personally. Allowing duplicates decreases the code size, so I'm all for the switch personally.

Unsquashed (https://github.com/kallewoof/bitcoin/commit/8871340d2aeb5e2e7888430eee0ed358c67c3030 is previous head): https://github.com/kallewoof/bitcoin/tree/listsinceblock-include-lost-txs-unsquashed
- https://github.com/kallewoof/bitcoin/commit/c7adf4aee94c5663826af4b277bca632e4f5cf31 addresses nits
- https://github.com/kallewoof/bitcoin/commit/12831108c0abdc088814a3c3d65c5e75387f05b0 and https://github.com/kallewoof/bitcoin/commit/2475789299304366c86f790fb05de27edea679b6 (test tweak) fixes duplicates; if people oppose this change I can drop these commits easily enough

luke-jr commented:
Do you plan to remove the `allow_partial` so it always throws an exception?

kallewoof commented:
@luke-jr I would love to, personally. I wasn't sure if it was the agreed approach, in the end, but more people have agreed about throwing than not, so I'm going to remove it for now.

kallewoof commented on test/functional/listsinceblock.py:
```diff
@@ -77,5 +77,183 @@ def run_test (self):
+
+        # but it should not include 'removed' if include_removed=false
+        lsbres2 = self.nodes[0].listsinceblock(lastblockhash, 1, False, False)
+        assert_equal('removed' in lsbres2, False)
```
I somehow skipped over this one. Fixed in squashed, unsquashed commit is https://github.com/kallewoof/bitcoin/commit/72ec1eaeda81c63a3e22fe0985ec02537082c9eb

jnewbery commented:
Thanks @kallewoof - I don't really mind too much either way. If I were coding this from scratch, I would prefer to show duplicates in removed, but I'm happy to go along with the consensus and since you've already put in the code to hide duplicates, that's also fine.

kallewoof commented:
@jnewbery Duplicates **are** shown in `removed` now. What I removed was a different feature, where a user could pass in an 'allow partial' flag which would prevent the node from throwing an exception if a block couldn't be read (useful for pruned nodes), but several people preferred always throwing.

kallewoof commented:
Realized I had missed one `assert_equal(x, false) -> assert not x`.

Unsquashed history:
* Mar 23: [1](https://github.com/kallewoof/bitcoin/commit/1a1398a11d9c4ddb1c1dfcff3251e6ecaa161f41) → [2](https://github.com/kallewoof/bitcoin/commit/f7405744c45cb7ea614b1896c2ac437374130803) → [3⊱1](https://github.com/kallewoof/bitcoin/commit/bdd743275d3141abbd3ed21843903df1a6891daa) → [~~4~~](https://github.com/kallewoof/bitcoin/commit/021e90136174e6acd77f86c030e593740a407fd0) → [~~5~~](https://github.com/kallewoof/bitcoin/commit/f959f77ddbe8355a6636adff40c3e3e0620e052b) → [6⊱1](https://github.com/kallewoof/bitcoin/commit/53322778ea82db4cd4c8ac64758e55b24f1f7c38) → [7⊱1](https://github.com/kallewoof/bitcoin/commit/8871340d2aeb5e2e7888430eee0ed358c67c3030) 
* Apr 13: → [8⊱2](https://github.com/kallewoof/bitcoin/commit/c7adf4aee94c5663826af4b277bca632e4f5cf31) → [9⊱1](https://github.com/kallewoof/bitcoin/commit/12831108c0abdc088814a3c3d65c5e75387f05b0) → [10⊱2](https://github.com/kallewoof/bitcoin/commit/2475789299304366c86f790fb05de27edea679b6) → [11⊱1](https://github.com/kallewoof/bitcoin/commit/501dbfeabe83c902c0aa677a15681932e2e1bfed)
* Apr 17: → [12⊱2](https://github.com/kallewoof/bitcoin/commit/72ec1eaeda81c63a3e22fe0985ec02537082c9eb)
* Apr 20: → [13⊱2](https://github.com/kallewoof/bitcoin/commit/55acfddc1271cd9dee596a80cfd88caac97dac41)

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1792,6 +1810,8 @@ UniValue listsinceblock(const JSONRPCRequest& request)
         filter = filter | ISMINE_WATCH_ONLY;
     }
 
+    bool include_removed = (request.params.size() < 4 || request.params[3].get_bool());
```
When adding RPCs or arguments, good advice is to treat null arguments the same as missing arguments: https://github.com/bitcoin/bitcoin/pull/10143#discussion_r112140887

I think that means this line would become:

`bool include_removed = (request.params.size() < 4 || request.params[3].isNull() || request.params[3].get_bool());`

jnewbery commented on test/functional/listsinceblock.py:
```diff
@@ -14,7 +14,12 @@ def __init__(self):
 
-    def run_test (self):
+    def run_test(self):
+        self.test_reorg()
```
Now that you've split this into sub-tests, can you move the lines:

```python
self.nodes[2].generate(101)
self.sync_all()
```

into the `run_test()` function before calling `test_reorg()`. It's best if sub-tests have no shared state between them. You should be able to re-order or skip tests without breaking them, and generating the chain in `test_reorg()` breaks that assumption.

We want sub-tests to be as independent as possible so if they break it's easier to debug where the problem is.

kallewoof replied:
Done - but some tweaking is necessary to make them truly independent of order (e.g. instead of asserting on getbalance()s, should just grab balance and compare). Will not do that work in this PR, but will try to address it asap unless someone else gets to it before me.

jnewbery commented on test/functional/listsinceblock.py:
```diff
@@ -75,7 +80,176 @@ def run_test (self):
+            self.nodes[3].getnewaddress(): 1,
+            self.nodes[2].getnewaddress(): change,
+        }
+        txid2 = self.nodes[2].sendrawtransaction(
```
`txid2` is unused so you don't need to assign it here.

jnewbery commented on test/functional/listsinceblock.py:
```diff
@@ -75,7 +80,176 @@ def run_test (self):
+        assert any(tx['txid'] == txid1 for tx in lsbres['removed'])
+
+        # but it should not include 'removed' if include_removed=false
+        lsbres2 = self.nodes[0].listsinceblock(lastblockhash, 1, False, False)
```
consider using named arguments here so it's clear to the reader what the arguments are for (you can also omit the optional `target_confirmations` and `include_watchonly` arguments)

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1756,7 +1773,8 @@ UniValue listsinceblock(const JSONRPCRequest& request)
     LOCK2(cs_main, pwallet->cs_wallet);
 
-    const CBlockIndex *pindex = NULL;
+    const CBlockIndex* pindex = NULL;
```
These variables might warrant their own comments now, since it's not immediately obvious what they're doing:

pindex - transactions from this block onwards should be included in the result. If the specified block was not in the main chain, pindex is the block where the chain forked
paltindex - used to count back from the specified block to the fork point to collect transactions for the removed array.

You can probably come up with better wording.

jnewbery reviewed (approved):
Looks good. A few nits but otherwise ACK https://github.com/bitcoin/bitcoin/pull/9622/commits/a8c56bf7b765910fb049373655c2823c7f8b6971

kallewoof commented:
@jnewbery Thanks for feedback! I think I addressed all of the stuff you mentioned. Compare [14⊱1](https://github.com/kallewoof/bitcoin/commit/fc23dfaa37845327f383d890b83ff63b6ac9bb73) & [15⊱2](https://github.com/kallewoof/bitcoin/commit/b7dd458485fdbceddbd5776a0fbeeee825aa70c1).

[[...]](https://github.com/kallewoof/bitcoin/commits/listsinceblock-include-lost-txs-unsquashed):
* Apr 20: → [13⊱2](https://github.com/kallewoof/bitcoin/commit/55acfddc1271cd9dee596a80cfd88caac97dac41)
* Apr 21: → [14⊱1](https://github.com/kallewoof/bitcoin/commit/fc23dfaa37845327f383d890b83ff63b6ac9bb73) → [15⊱2](https://github.com/kallewoof/bitcoin/commit/b7dd458485fdbceddbd5776a0fbeeee825aa70c1)

jnewbery commented:
Looks good. ReACK https://github.com/bitcoin/bitcoin/pull/9622/commits/4578a21bc29768e11f56be600b141d81447db7a6

kallewoof commented:
Rebased due to conflicts.

jnewbery commented:
> Edit: removed some irrelevant asserts that started causing errors in tests. (Not sure why a clean chain with generate(101) would yield 5100 BTC instead of 50, but no matter.)

Looks like you've reintroduced the lines:
```python
        self.nodes[2].generate(101)
        self.sync_all()
```
into the `test_reorg()` function, which means you're now generating 202 blocks (so node 2 now has 102 spendable coinbases = 5100 BTC).

I think this is just a bad rebase.

kallewoof commented:
@jnewbery Ahh, thanks. Yeah, that is obviously it. I am just going to remove the double generate in test_reorg.

sipa commented:
utACK 5d352044ca7f29f992b4de46d461bccc86326ba1 with a few nits. I did not review the tests.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1727,19 +1744,20 @@ UniValue listsinceblock(const JSONRPCRequest& request)
     isminefilter filter = ISMINE_SPENDABLE;
 
-    if (request.params.size() > 0)
+    if (request.params.size() > 0 && !request.params[0].isNull())
```
Oh, I didn't know that. That makes things cleaner, thanks.

kallewoof commented:
@sipa Thanks for the review. I've addressed your nits in https://github.com/kallewoof/bitcoin/commit/3a10e7c49108dde97d999f46bbb36ce905887a0e

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1795,35 +1811,57 @@ UniValue listsinceblock(const JSONRPCRequest& request)
+                // Use -depth as minDepth parameter to ListTransactions to prevent incoming
+                // transactions from being filtered. These transactions have negative
+                // confirmations, but always greater than -depth.
+                ListTransactions(pwallet, pwallet->mapWallet[tx->GetHash()], "*", -depth, true, removed, filter);
```
Hmm, technically -depth isnt sufficient here. In the rare case of a reorg-to-lower-block-height this would be insufficient. You already got the block from the chain, just call it - 1000000000 or so.

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1761,7 +1776,10 @@ UniValue listsinceblock(const JSONRPCRequest& request)
             "    \"to\": \"...\",            (string) If a comment to is associated with the transaction.\n"
-             "  ],\n"
+            "  ],\n"
+            "  \"removed\": [\n"
```
Can we add something to note that transactions which were re-added are included here anyway, and may have, at that point, positive confirmations value in this array?

TheBlueMatt commented on test/functional/listsinceblock.py:
```diff
@@ -73,7 +73,171 @@ def run_test (self):
+        Asserted:
+
+        1. tx1 is listed in listsinceblock.
+        2. It is not included in 'removed' because it was not removed.
```
This is wrong (and the check and later comment contradict this).

TheBlueMatt commented on test/functional/listsinceblock.py:
```diff
@@ -73,7 +73,171 @@ def run_test (self):
+        assert any(tx['txid'] == txid1 for tx in lsbres['removed'])
+
+        # find transaction and ensure confirmations is valid
+        for tx in lsbres['transactions']:
```
Can you duplicate this loop for "removed", noting that the tx listed in "removed" should *also* have a confirmations of 2.

kallewoof commented:
@TheBlueMatt Thanks for the review! I've addressed all of your nits, I believe. The unsquashed changes are in https://github.com/kallewoof/bitcoin/commit/5166d0e4f819ab4b430103fa350f8f882f1b1e2f and https://github.com/kallewoof/bitcoin/commit/299c00c56b4361c46a2dd9cc2fa4d4beb8bc2250.

TheBlueMatt commented:
I think this maybe missed 15, sadly. Its a nice change, but not a bugfix.

kallewoof commented:
That's disheartening, but ah well.

TheBlueMatt commented:
utACK d6115c21a790e139bb45ec6dee421a976bc21eaf. Yea, Core has been in a near-constant state of growing pains for some time it seems. Recently its grown active enough to be just beyond the ability of any individual to keep up with everything, and so things occasionally move slower than they should :(. Anyway, up to @laanwj, this isnt a bugfix, but its pretty clean so maybe he's still willing to pull it for 15.

kallewoof commented:
I'd say let's just push it to 16 milestone. I'll be noisy about it this time. :)

TheBlueMatt commented:
Heh, yes, if things have been sitting for a while definitely be noisy!

On July 19, 2017 7:12:08 PM EDT, kallewoof <notifications@github.com> wrote:
>I'd say let's just push it to 16 milestone. I'll be noisy about it this
>time. :)

laanwj commented:
Needs rebase.

> I'd say let's just push it to 16 milestone. I'll be noisy about it this time. :)

Up to you, though given how long this has been open and how much interest and review this has I don't have a particular problem with merging this into 0.15 still.

kallewoof commented:
@laanwj Rebased. And I'd *love* to have this in 15, personally. I just don't wanna rush anything for the sake of the PR being old.

sipa commented:
utACK
